[stable/traefik] Added support for node and namespace selectors

electroma opened the pull request:
Minor features in traefik chart: 
-support for "namespaces" filter value
-support for "labelSelector" value value

This feature may be useful in case you need to have more than one instance of traefik-based ingress controllers.

electroma committed:
Added: support for node and namespace selectors

electroma committed:
Added: docs

electroma committed:
Minor: NL

lachie83 commented:
cc @krancour

krancour commented:
Mostly looks good, but I'd like to test manually.

krancour commented on stable/traefik/templates/configmap.yaml:
```diff
@@ -34,6 +34,12 @@ data:
+    {{- end}}
+    {{- if .Values.kubernetes.labelSelector }}
+    labelselector = {{ .Values.kubernetes.labelSelector | quote }}
+    {{- end}}
```
It would be a good idea to wrap everything from lines 37 to 42 in:

```
{{- if .Values.kubernetes}}
{{- end}}
```

This is necessary to accommodate the scenario where `values.yaml` either does not contain `kubernetes` or contains something like this:

```
kubernetes:
  # namespaces: ~
  # labelSelector: ~
```

In the former scenario (`kubernetes` is absent), `.Values.kubernetes` evaluates as `nil` and template processing will fail because a `nil` doesn't have a `namespaces` or `labelSelector` field. In the latter case (the sample above), `kubernetes` evaluates as an empty string and will have the same problem-- neither of those two fields will exist.

We learned this while working on / evaluating #414.

electroma replied:
Fixed 
@krancour thank you for spotting it!

krancour replied:
Thanks for fixing. This looks good and I will manually test by eod on Monday.

electroma committed:
Fixed: extra check for "kubernetes" key presence, thanks to @krancour

viglesiasce commented:
@k8s-bot ok to test

krancour commented on stable/traefik/values.yaml:
```diff
@@ -26,3 +26,6 @@ dashboard:
   enabled: true
+kubernetes:
+  namespaces: ~
+  labelSelector: ~
```
What's the meaning and effective result of the `~` here? I inferred it means "all," but I can't find in the Traefik docs where it says that. I think that rather than set these to `~` (whose meaning isn't clear), it might be better to leave these commented (and therefore undefined) by default. It will be more consistent with other pending PRs.

electroma replied:
The effective result is  "watch all namespaces with no filter". I've commented it out and removed "~"

krancour commented on stable/traefik/templates/service.yaml:
```diff
@@ -14,5 +14,7 @@ spec:
   ports:
   - port: 80
     name: http
+{{- if .Values.ssl.enabled }}
```
Can we take this guard out of this PR? I agree it probably belongs, but it's unrelated to everything else this PR is trying to accomplish. Smaller, more tightly-scoped PRs are more tractable.

electroma replied:
removed it

electroma committed:
Fixed comments from @krancour:
-removed condition on service https port
-ingress filter is commented out by default

krancour commented on stable/traefik/README.md:
```diff
@@ -87,7 +87,9 @@ The following tables lists the configurable parameters of the Traefik chart and
 | `dashboard.domain`              | Domain for the Traefik dashboard                                     | `traefik.example.com`                     |
-| `gzip.enabled`                  | Whether to use gzip compression                                      | `true`                     |
+| `gzip.enabled`                  | Whether to use gzip compression                                      | `true`                                    |
+| `kubernetes.namespaces`         | List of Kubernetes namespaces to watch (e. g. `["default"]`)         |                                           |
```
Ok... this has to change. I didn't read this carefully before. I just saw the word "list" and naturally did this:

```
kubernetes:
  namespaces:
  - foo
  - bar
```

^ That's the native YAML way of expressing a list. Requiring a format like `["default"]` feels very awkward because it's mixing TOML syntax with YAML syntax.

electroma replied:
Fixed and tested with empty, one and several values

krancour commented on stable/traefik/templates/configmap.yaml:
```diff
@@ -34,6 +34,14 @@ data:
     [kubernetes]
+    {{- if .Values.kubernetes}}
+      {{- if .Values.kubernetes.namespaces }}
+    namespaces = {{ .Values.kubernetes.namespaces }}
```
Should range over the YAML list here to create a TOML list as the value for `namespaces`.

electroma committed:
Fixed configuration: accept list of namespaces instead of TOML string

krancour commented:
LGTM

Commit:
bump chart version

prydonius commented:
rebased and bumped @viglesiasce
